MODE=gnu-ld produces broken .c files

When trying to build the included python example:

../../w2c2/w2c2 -d gnu-ld -f 100 python.wasm python.c
w2c2: skipping custom section 'producers' (size 85)
w2c2: skipping custom section 'target_features' (size 28)
make -fSubmake CC="cc" CFLAGS="-I../../w2c2 -O0 -g" LDFLAGS="-lm -L../../wasi -lw2c2wasi" OBJS="datasegments.o"
cc -I.. -I../../w2c2 -O0 -g -c python.c -o python.o
ld -r -b binary -o datasegments.o datasegments
python.c:8:17: error: invalid suffix "const" on integer constant
    8 | const U8* d0=ds+0const U8* d1=ds+5const U8* d2=ds+18const U8* d3=ds+29const U8* d4=ds+48const U8* d5=ds+83const U8* d6=ds+88const U8* d7=ds+169const U8* d8=ds+176const U8* d9=ds+177const U8* d10=ds+190const U8* d11=ds+191const U8* d12=ds+209const U8* d13=ds+218const U8* d14=ds+228const U8* d15=ds+239const U8* d16=ds+258const U8* d17=ds+293const U8* d18=ds+297const U8* d19=ds+417const U8* d20=ds+419const U8* d21=ds+420const U8* d22=ds+438const U8* d23=ds+447const U8* d24=ds+453const U8* d25=ds+464const U8* d26=ds+483const U8* d27=ds+518const U8* d28=ds+523const U8* d29=ds+560const U8* d30=ds+637const U8* d31=ds+639const U8* d32=ds+640const U8* d33=ds+658const U8* d34=ds+667const U8* d35=ds+677const U8* d36=ds+680constU8* d37=ds+689const U8* d38=ds+692const U8* d39=ds+708const U8* d40=ds+713const U8* 

Confirmed on Ubuntu and macOS.

--- a/w2c2/c.c
+++ b/w2c2/c.c
@@ -5117,13 +5117,8 @@ wasmCWriteDataSegments(
 
     {
         U32 dataSegmentIndex = 0;
-        U64 byteOffset = 0;
         for (; dataSegmentIndex < dataSegmentCount; dataSegmentIndex++) {
             const WasmDataSegment dataSegment = module->dataSegments.dataSegments[dataSegmentIndex];
-
-            if (!dataSegment.bytes.length) {
-                continue;
-            }
 
             switch (mode) {
                 case wasmDataSegmentModeArrays: {
@@ -5155,15 +5150,10 @@ wasmCWriteDataSegments(
                 case wasmDataSegmentModeGNULD:
                 case wasmDataSegmentModeSectcreate1:
                 case wasmDataSegmentModeSectcreate2: {
-                    fputs("const U8* ", file);
+                    fputs("U8* ", file);
                     /* TODO: add support for multiple modules */
                     wasmCWriteFileDataSegmentName(file, dataSegmentIndex);
-                    if (pretty) {
-                        fprintf(file, " = ds + %llu", byteOffset);
-                    } else {
-                        fprintf(file, "=ds+%llu", byteOffset);
-                    }
-                    byteOffset += dataSegment.bytes.length;
+                    fputs(";\n", file);
                     break;
                 }
                 default: {
@@ -5386,39 +5376,52 @@ wasmCWriteInitMemories(
         {
             const U32 dataSegmentCount = module->dataSegments.count;
             U32 dataSegmentIndex = 0;
+            U64 byteOffset = 0;
             for (; dataSegmentIndex < dataSegmentCount; dataSegmentIndex++) {
                 const WasmDataSegment dataSegment = module->dataSegments.dataSegments[dataSegmentIndex];
                 const size_t dataSegmentLength = dataSegment.bytes.length;
                 const Buffer code = dataSegment.offset;
 
-                if (!dataSegmentLength) {
-                    continue;
+                switch (dataSegmentMode) {
+                    case wasmDataSegmentModeGNULD:
+                    case wasmDataSegmentModeSectcreate1:
+                    case wasmDataSegmentModeSectcreate2: {
+                        /* TODO: add support for multiple modules */
+                        wasmCWriteFileDataSegmentName(file, dataSegmentIndex);
+                        if (pretty) {
+                            fprintf(file, " = ds + %llu", byteOffset);
+                        } else {
+                            fprintf(file, "=ds+%llu", byteOffset);
+                        }
+                        fputs(";\n", file);
+                        break;
+                    }
                 }
 
-                /* Skip passive segments */
-                if (code.data == NULL) {
-                    continue;
+                /* Load active segments */
+                if (code.data != NULL) {
+                    if (pretty) {
+                        fputs(indentation, file);
+                    }
+                    fputs("LOAD_DATA(", file);
+                    wasmCWriteFileMemoryUse(
+                            file,
+                            module,
+                            dataSegment.memoryIndex,
+                            NULL,
+                            false
+                    );
+                    fputs(", ", file);
+                    MUST (stringBuilderReset(&stringBuilder))
+                    MUST (wasmCWriteConstantExpr(&stringBuilder, module, code))
+                    fputs(stringBuilder.string, file);
+                    fputs(", ", file);
+                    /* TODO: add support for multiple modules */
+                    wasmCWriteFileDataSegmentName(file, dataSegmentIndex);
+                    fprintf(file, ", %lu);\n", (unsigned long) dataSegmentLength);
                 }
 
-                if (pretty) {
-                    fputs(indentation, file);
-                }
-                fputs("LOAD_DATA(", file);
-                wasmCWriteFileMemoryUse(
-                    file,
-                    module,
-                    dataSegment.memoryIndex,
-                    NULL,
-                    false
-                );
-                fputs(", ", file);
-                MUST (stringBuilderReset(&stringBuilder))
-                MUST (wasmCWriteConstantExpr(&stringBuilder, module, code))
-                fputs(stringBuilder.string, file);
-                fputs(", ", file);
-                /* TODO: add support for multiple modules */
-                wasmCWriteFileDataSegmentName(file, dataSegmentIndex);
-                fprintf(file, ", %lu);\n", (unsigned long) dataSegmentLength);
+                byteOffset += dataSegment.bytes.length;
             }
         }
 

